{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**13 June 2016**\n",
    "\n",
    "For yeast, files under directories `short-single-dx` where x = 50, 100, 200 created with commands:\n",
    "```\n",
    "$ PREFIX=/global/scratch/seanla/Data/yeast\n",
    "$ OUTPUTDIR=$PREFIX/short-single-d200\n",
    "$ OUTPUT=$OUTPUTDIR/yeast-short-single-d200\n",
    "$ cd /home/seanla/Software/art_bin_GreatSmokyMountains/\n",
    "$ mkdir -p $OUTPUTDIR\n",
    "$ ./art_illumina -s -sam -i $PREFIX/saccharomyces-cerevisiae chromosome1_sequence.fasta -l 100 -f x -o $OUTPUT\n",
    "```\n",
    "where x = 50, 100, 200."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For E.coli, files under directories `short-single-dx` where x = 50, 100, 200 created with commands:\n",
    "```\n",
    "$ PREFIX=/global/scratch/seanla/Data/ecoli\n",
    "$ OUTPUTDIR=$PREFIX/short-single-d200\n",
    "$ OUTPUT=$OUTPUTDIR/ecoli-short-single-d200\n",
    "$ cd /home/seanla/Software/art_bin_GreatSmokyMountains/\n",
    "$ mkdir -p $OUTPUTDIR\n",
    "$ ./art_illumina -s -sam -i $PREFIX/escherichia-coli_reference.fasta -l 100 -f x -o $OUTPUT\n",
    "```\n",
    "where x = 50, 100, 200\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**16 June 2016**\n",
    "\n",
    "Simulated long read files for E.coli with coverage of 10, 20, 50 and 75 were generated by SimLoRD using a shell/PBS script with the commands\n",
    "\n",
    "```\n",
    "PREFIX=/global/scratch/seanla/Data/ecoli\n",
    "PATH=/home/seanla/Projects/hcstats\n",
    "REF=$PREFIX/escherichia-coli_reference.fasta\n",
    "FASTQ=$PREFIX/SRR1284073.fastq\n",
    "\n",
    "\n",
    "COVERAGE=x\n",
    "READS=$(/global/software/python/python272/bin/python $PATH/reads4coverage.py -c $COVERAGE -i /global/scratch/seanla/Data/ecoli/SRR1284073.fastq -r /global/scratch/seanla/Data/ecoli/escherichia-coli_reference.fasta)\n",
    "OUTPUT=\"$PREFIX/simlord/long-d$COVERAGE/ecoli-long-d$COVERAGE\"\n",
    "\n",
    "/bin/mkdir -p $OUTPUT\n",
    "\n",
    "echo \"Coverage: $COVERAGE\"\n",
    "echo \"Reads: $READS\"\n",
    "\n",
    "/home/seanla/Software/anaconda2/envs/simlord/bin/simlord -n $READS -sf $FASTQ -rr $REF $OUTPUT\n",
    "```\n",
    "\n",
    "Where `x` was set to 10, 20, 50 and 75 respectively."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Simulated long read files for yeast with coverage of 10, 20, 50 and 75 were generated by SimLoRD using a shell/PBS script with the following commands:\n",
    "\n",
    "```\n",
    "# Paths to executables\n",
    "python=/global/software/python/python272/bin/python\n",
    "reads4coverage=/home/seanla/Projects/hcstats/reads4coverage.py\n",
    "simlord=/home/seanla/Software/anaconda2/envs/simlord/bin/simlord\n",
    "\n",
    "# Paths to data\n",
    "data=/global/scratch/seanla/Data/yeast\n",
    "ref=$data/saccharomyces-cerevisiae-chromosome1_sequence.fasta\n",
    "fastq=$data/SRR1284662.fastq\n",
    "\n",
    "# Perform simlord for coverage 10, 20, 50, 75\n",
    "\n",
    "# ------------------ Coverage x -----------------------------\n",
    "# Input parameters for simlord\n",
    "coverage=x\n",
    "# Find number of reads required for coverage\n",
    "reads=$($python $reads4coverage -c $coverage -i $fastq -r $ref)\n",
    "# Output directory path\n",
    "output=\"$data/simlord/long-d$coverage/ecoli-long-d$coverage\"\n",
    "\n",
    "mkdir -p $output\n",
    "\n",
    "echo \"Coverage: $coverage\"\n",
    "echo \"Reads: $reads\"\n",
    "\n",
    "$simlord -n $reads -sf $fastq -rr $ref $output\n",
    "```\n",
    "where `x` is 10, 20, 50 and 75."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "** 20 June 2016 **\n",
    "\n",
    "Paired-end short reads of coverage 50, 100 and 200 for ecoli and yeast were generated with the following commands:\n",
    "\n",
    "```\n",
    "prefix=/global/scratch/seanla/Data/ecoli\n",
    "outputdir=$prefix/art/short-paired-d200\n",
    "output=$outputdir/ecoli-short-paired-d200\n",
    "ref=$prefix/escherichia-coli_reference.fasta\n",
    "mkdir -p $outputdir\n",
    "cd /home/seanla/Software/art_bin_GreatSmokyMountains/\n",
    "./art_illumina -p -i $ref -l 100 -f x -m 300 -s 25 -o $output\n",
    "```\n",
    "\n",
    "```\n",
    "prefix=/global/scratch/seanla/Data/yeast\n",
    "outputdir=$prefix/art/short-paired-d100\n",
    "output=$outputdir/yeast-short-paired-d100\n",
    "ref=$prefix/accharomyces-cerevisiae-chromosome1_sequence.fasta\n",
    "mkdir -p $outputdir\n",
    "cd /home/seanla/Software/art_bin_GreatSmokyMountains/\n",
    "./art_illumina -p -i $ref -l 100 -f x -m 300 -s 25 -o $output\n",
    "```\n",
    "\n",
    "Where the `-l 100` indicates a short read length of 100, `-f x` indicates a fold coverage of `x` (of which `x` was set to 50, 100 and 200), `-m 300` indicates a mean fragment length of 300 and `-s 25` indicates a standard deviation of fragment length of 25."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "**21 June 2016**\n",
    "LoRDeC, Jabba and proovread were performed on short reads of coverage 50, 100 and 200 and long reads of coverage 10, 20, 50 and 75.\n",
    "\n",
    "For LoRDeC, the following parameters were used:\n",
    "- kmer length of 19\n",
    "- error rate of 0.4\n",
    "- number of target k-mers: 5\n",
    "- max branch exploration of 200\n",
    "\n",
    "For Jabba, the following parameters were used:\n",
    "- The de Bruijn graph was created with Brownie with a k-mer length of 45 (graphCorrection mode)\n",
    "- Thread usage of 6\n",
    "- Minimum size of MEM of 20\n",
    "- k-mer size of 45\n",
    "\n",
    "For proovread, preprocessing was performed as follows:\n",
    "- bwa-proovread was used to map the short reads onto the long reads\n",
    "- the resulting SAM file was sorted using samtools\n",
    "- For E.coli runs 50Sx50L, 50Sx75L, 100Sx50L, 100Sx75L, 200Sx10L, 200Sx20L, 200Sx50L, 200Sx75L, the parameter \"long read quality offset value\" could not be automatically determined and so the corrections failed. In future runs this value must be inputted manually.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
