{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Real Long Read Data #\n",
    "For this experiment, the following true E.coli genomic data were used:\n",
    "#### Reference genome ####\n",
    "* [_Escherichia coli_ str. K-12 substr. MG1655, complete genome](http://www.ncbi.nlm.nih.gov/nuccore/NC_000913)\n",
    "\n",
    "#### PacBio Long Reads ####\n",
    "* [_E.coli_ K12 MG1655 CLR Reads from Pacific Biosciences DevNet](https://github.com/PacificBiosciences/DevNet/wiki/E%20coli%20K12%20MG1655%20Hybrid%20Assembly)\n",
    "* [_E.coli_ K12 MG1655 CCS Reads from Pacific Biosciences Devnet](https://github.com/PacificBiosciences/DevNet/wiki/E%20coli%20K12%20MG1655%20Hybrid%20Assembly)\n",
    "* [Pacific Biosciences Model Organism Genome Sequencing of _E.coli_ P5C3 - SRR1284073](http://www.ncbi.nlm.nih.gov/sra/?term=SRR1284073)\n",
    "\n",
    "## Analyses ##\n",
    "Length distribution histograms and length-accuracy scatterplots were created for each dataset and mean identities were also found.\n",
    "\n",
    "#### Definitions ####\n",
    "Given an alignment between reference sequence A and read B, the _identity_ of B is defined as the number of positions in the alignment such that the bases of A and B are equal.\n",
    "\n",
    "### Mean Identities ###\n",
    "\n",
    "|  Dataset   | Mean Identity |\n",
    "| :-------:  | :-----------: |\n",
    "| CLR        | 1969          |\n",
    "| CCS        | 692           |\n",
    "| SRR1284073 | 5325          |\n",
    "\n",
    "### Length Distributions ###\n",
    "\n",
    "#### CLR ####\n",
    "![CLR Length Distribution](ecoli_real/length_distributions/PacBio_10kb_CLR_length_dist.png)\n",
    "\n",
    "#### CCS ####\n",
    "![CCS Length Distribution](ecoli_real/length_distributions/PacBio_2kb_CCS_500bp_length_dist.png)\n",
    "\n",
    "#### SRR1284073 ####\n",
    "![SRR1284073 Length Distribution](ecoli_real/length_distributions/SRR1284073_length_dist.png)\n",
    "\n",
    "### Length-Identity Scatterplots ###\n",
    "\n",
    "#### CLR ####\n",
    "![CLR Length-Accuracy Scatterplots](ecoli_real/length_identity_scatters/PacBio_10kb_CLR_length_identity_scatter.png)\n",
    "\n",
    "#### CCS ####\n",
    "![CCS Length-Accuracy Scatterplots](ecoli_real/length_identity_scatters/PacBio_2kb_CCS_length_identity_scatter.png)\n",
    "\n",
    "#### SRR1284073 ####\n",
    "![SRR1284073 Length-Accuracy Scatterplots](ecoli_real/length_identity_scatters/SRR1284073_length_identity_scatter.png)\n",
    "\n",
    "# Simulated Data #\n",
    "PacBio long reads were simulated using SimLoRD. Two types of reads were simulated: CLR reads (e.g. maximum number of passes over each each read was set to 1) and CCS reads (e.g. maximum number of passes over each read was set to 40). For each type of read, datasets of long reads with coverages 10, 20, 50 and 75 were simulated.\n",
    "\n",
    "## CLR Analyses ##\n",
    "Length distribution histograms, length-accuracy scatterplots, and length - error rate scatter plots were created for each dataset and mean identities were also found.\n",
    "\n",
    "#### Definitions ####\n",
    "Given an alignment between reference sequence A and read B, the _identity_ of B is defined as the number of positions in the alignment such that the bases of A and B are equal.\n",
    "\n",
    "Given an alignment between reference sequence A and read B, the _error rate_ of B is the number of positions in the alignment such that the bases of A and B are not equal.\n",
    "\n",
    "Given a set of SimLoRD reads S, the _total error rate_ of S is defined as the total number of erroneous bases in S divided by the total number of bases in S. This number is given a priori in the SimLoRD FASTQ files.\n",
    "\n",
    "### Mean Identities ###\n",
    "\n",
    "|  Coverage   | Mean Identity |\n",
    "| :--------:  | :-----------: |\n",
    "| 10          | 2884          |\n",
    "| 20          | 2873          |\n",
    "| 50          | 2897          |\n",
    "| 75          | 2879          |\n",
    "\n",
    "### Error Rate ###\n",
    "\n",
    "| Coverage | Error Rate |\n",
    "| :------: | :--------: |\n",
    "| 10       | 0.165      |\n",
    "| 20       | 0.165      |\n",
    "| 50       | 0.165      |\n",
    "| 75       | 0.165      |\n",
    "\n",
    "### Length Distributions ###\n",
    "\n",
    "#### Coverage 10 ####\n",
    "![Coverage 10 Length Distribution](ecoli_clr/length_distributions/long-d10.png)\n",
    "\n",
    "#### Coverage 20 ####\n",
    "![Coverage 20 Length Distribution](ecoli_clr/length_distributions/long-d20.png)\n",
    "\n",
    "#### Coverage 50 ####\n",
    "![Coverage 50 Length Distribution](ecoli_clr/length_distributions/long-d50.png)\n",
    "\n",
    "#### Coverage 75 ####\n",
    "![Coverage 75 Length Distribution](ecoli_clr/length_distributions/long-d75.png)\n",
    "\n",
    "### Length - Identity Scatterplots ###\n",
    "\n",
    "#### Coverage 10 ####\n",
    "![Coverage 10 Length Identity Scatterplot](ecoli_clr/length_identity_scatters/ecoli-long-d10_length_identity_scatter.png)\n",
    "\n",
    "#### Coverage 20 ####\n",
    "![Coverage 20 Length Identity Scatterplot](ecoli_clr/length_identity_scatters/ecoli-long-d20_length_identity_scatter.png)\n",
    "\n",
    "#### Coverage 50 ####\n",
    "![Coverage 50 Length Identity Scatterplot](ecoli_clr/length_identity_scatters/ecoli-long-d50_length_identity_scatter.png)\n",
    "\n",
    "#### Coverage 75 ####\n",
    "![Coverage 75 Length Identity Scatterplot](ecoli_clr/length_identity_scatters/ecoli-long-d75_length_identity_scatter.png)\n",
    "\n",
    "### Length - Error Rate Scatterplots ###\n",
    "\n",
    "#### Coverage 10 ####\n",
    "![Coverage 10 Length - Error Rate Scatter](ecoli_clr/error_rate_scatter/ecoli-long-d10_error_rate_scatter.png)\n",
    "\n",
    "#### Coverage 20 ####\n",
    "![Coverage 20 Length - Error Rate Scatter](ecoli_clr/error_rate_scatter/ecoli-long-d20_error_rate_scatter.png)\n",
    "\n",
    "#### Coverage 50 ####\n",
    "![Coverage 50 Length - Error Rate Scatter](ecoli_clr/error_rate_scatter/ecoli-long-d50_error_rate_scatter.png)\n",
    "\n",
    "#### Coverage 75 ####\n",
    "![Coverage 75 Length - Error Rate Scatter](ecoli_clr/error_rate_scatter/ecoli-long-d75_error_rate_scatter.png)\n",
    "\n",
    "## CCS Analysis ##\n",
    "Length distribution histograms, length-accuracy scatterplots, and length - true error rate scatter plots were created for each dataset and mean identities were also found.\n",
    "\n",
    "#### Definitions ####\n",
    "Given an alignment between reference sequence A and read B, the _identity_ of B is defined as the number of positions in the alignment such that the bases of A and B are equal.\n",
    "\n",
    "Given an alignment between reference sequence A and read B, the _error rate_ of B is the number of positions in the alignment such that the bases of A and B are not equal.\n",
    "\n",
    "Given a set of SimLoRD reads S, the _total error rate_ of S is defined as the total number of erroneous bases in S divided by the total number of bases in S. This number is given a priori in the SimLoRD FASTQ files.\n",
    "\n",
    "### Mean Identities ###\n",
    "\n",
    "|  Coverage   | Mean Identity |\n",
    "| :--------:  | :-----------: |\n",
    "| 10          | 2838          |\n",
    "| 20          | 2843          |\n",
    "| 50          | 2791          |\n",
    "| 75          | 2821          |\n",
    "\n",
    "### Error Rate ###\n",
    "\n",
    "| Coverage | Error Rate |\n",
    "| :------: | :--------: |\n",
    "| 10       | 0.096      |\n",
    "| 20       | 0.096      |\n",
    "| 50       | 0.095      |\n",
    "| 75       | 0.096      |\n",
    "\n",
    "### Length Distributions ###\n",
    "\n",
    "#### Coverage 10 ####\n",
    "![Coverage 10 Length Distribution](ecoli_ccs/length_distributions/long-d10.png)\n",
    "\n",
    "#### Coverage 20 ####\n",
    "![Coverage 20 Length Distribution](ecoli_ccs/length_distributions/long-d20.png)\n",
    "\n",
    "#### Coverage 50 ####\n",
    "![Coverage 50 Length Distribution](ecoli_ccs/length_distributions/long-d50.png)\n",
    "\n",
    "#### Coverage 75 ####\n",
    "![Coverage 75 Length Distribution](ecoli_ccs/length_distributions/long-d75.png)\n",
    "\n",
    "### Length - Error Rate Scatterplots ###\n",
    "\n",
    "#### Coverage 10 ####\n",
    "![Coverage 10 Length - Error Rate Scatter](ecoli_ccs/error_rate_scatter/ecoli_ccs-long-d10_error_rate_scatter.png)\n",
    "\n",
    "#### Coverage 20 ####\n",
    "![Coverage 20 Length - Error Rate Scatter](ecoli_ccs/error_rate_scatter/ecoli_ccs-long-d20_error_rate_scatter.png)\n",
    "\n",
    "#### Coverage 50 ####\n",
    "![Coverage 50 Length - Error Rate Scatter](ecoli_ccs/error_rate_scatter/ecoli_ccs-long-d50_error_rate_scatter.png)\n",
    "\n",
    "#### Coverage 75 ####\n",
    "![Coverage 75 Length - Error Rate Scatter](ecoli_ccs/error_rate_scatter/ecoli_ccs-long-d75_error_rate_scatter.png)\n",
    "\n",
    "### Length - Identity Scatterplots ###\n",
    "\n",
    "#### Coverage 10 ####\n",
    "![Coverage 10 Length Identity Scatterplot](ecoli_ccs/length_identity_scatters/ecoli-long-d10_length_identity_scatter.png)\n",
    "\n",
    "#### Coverage 20 ####\n",
    "![Coverage 20 Length Identity Scatterplot](ecoli_ccs/length_identity_scatters/ecoli-long-d20_length_identity_scatter.png)\n",
    "\n",
    "#### Coverage 50 ####\n",
    "![Coverage 50 Length Identity Scatterplot](ecoli_ccs/length_identity_scatters/ecoli-long-d50_length_identity_scatter.png)\n",
    "\n",
    "#### Coverage 75 ####\n",
    "![Coverage 75 Length Identity Scatterplot](ecoli_ccs/length_identity_scatters/ecoli-long-d75_length_identity_scatter.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [Root]",
   "language": "python",
   "name": "Python [Root]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
