{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Real Long Read Data #\n",
    "For this experiment, the following true E.coli genomic data were used:\n",
    "#### Reference genome ####\n",
    "* [_Escherichia coli_ str. K-12 substr. MG1655, complete genome](http://www.ncbi.nlm.nih.gov/nuccore/NC_000913)\n",
    "\n",
    "#### PacBio Long Reads ####\n",
    "* [_E.coli_ K12 MG1655 CLR Reads from Pacific Biosciences DevNet](https://github.com/PacificBiosciences/DevNet/wiki/E%20coli%20K12%20MG1655%20Hybrid%20Assembly)\n",
    "* [_E.coli_ K12 MG1655 CCS Reads from Pacific Biosciences Devnet](https://github.com/PacificBiosciences/DevNet/wiki/E%20coli%20K12%20MG1655%20Hybrid%20Assembly)\n",
    "* [Pacific Biosciences Model Organism Genome Sequencing of _E.coli_ P5C3 - SRR1284073](http://www.ncbi.nlm.nih.gov/sra/?term=SRR1284073)\n",
    "\n",
    "## Analyses ##\n",
    "Length distribution histograms and length-identity scatterplots were created for each dataset and mean identities were also found.\n",
    "\n",
    "To construct the length - identity scatterplots, the reads were mapped onto the reference genome using BLASR and the outputted SAM files were coverted to [Multiple Alignment Format](https://genome.ucsc.edu/FAQ/FAQformat.html#format5) for analyses.\n",
    "\n",
    "#### Definitions ####\n",
    "Given an alignment between reference sequence A and read B, the _identity_ of B is defined as the number of positions in the alignment such that the bases of A and B are equal.\n",
    "\n",
    "### Mean Identities ###\n",
    "\n",
    "|  Dataset   | Mean Identity |\n",
    "| :-------:  | :-----------: |\n",
    "| CLR        | 1969          |\n",
    "| CCS        | 692           |\n",
    "| SRR1284073 | 5325          |\n",
    "\n",
    "### Length Distributions ###\n",
    "\n",
    "#### CLR ####\n",
    "![CLR Length Distribution](ecoli_real/length_distributions/PacBio_10kb_CLR_length_dist.png)\n",
    "\n",
    "#### CCS ####\n",
    "![CCS Length Distribution](ecoli_real/length_distributions/PacBio_2kb_CCS_500bp_length_dist.png)\n",
    "\n",
    "#### SRR1284073 ####\n",
    "![SRR1284073 Length Distribution](ecoli_real/length_distributions/SRR1284073_length_dist.png)\n",
    "\n",
    "### Length-Identity Scatterplots ###\n",
    "\n",
    "#### CLR ####\n",
    "![CLR Length-Identity Scatterplots](ecoli_real/length_identity_scatters/PacBio_10kb_CLR_length_identity_scatter.png)\n",
    "\n",
    "#### CCS ####\n",
    "![CCS Length-Identity Scatterplots](ecoli_real/length_identity_scatters/PacBio_2kb_CCS_length_identity_scatter.png)\n",
    "\n",
    "#### SRR1284073 ####\n",
    "![SRR1284073 Length-Identity Scatterplots](ecoli_real/length_identity_scatters/SRR1284073_length_identity_scatter.png)\n",
    "\n",
    "# Simulated Data #\n",
    "PacBio long reads were simulated using SimLoRD. Two types of reads were simulated: CLR reads (e.g. maximum number of passes over each each read was set to 1) and CCS reads (e.g. maximum number of passes over each read was set to 40). For each type of read, datasets of long reads with coverages 10, 20, 50 and 75 were simulated.\n",
    "\n",
    "## Simulation of Short Reads ##\n",
    "Short reads were simulated using ART using the following command:\n",
    "\n",
    "```\n",
    "art-illumina -p -i $ref -l 100 -f $cov -m 300 -s 25 -o $outputPrefix\n",
    "```\n",
    "\n",
    "An explanation of the parameters used are as such:\n",
    "* `-l 100` indicates that the length of the reads simulated were 100.\n",
    "* `-f $cov` indicates that the fold of read coverage was set to `$cov`. `$cov` was then set to 50, 100 and 200.\n",
    "* `-m 300` indicates that the mean fragment size of paired end reads was set to 300.\n",
    "* `-s 25` indicates that the standard deviation of the fragment size of the reads was set to 25.\n",
    "\n",
    "## Simulation of Long Reads ##\n",
    "The simulation of long reads were done in two steps:\n",
    "1. Given the length of reads from a real PacBio FASTQ file, the number of reads required for a given coverage were found.\n",
    "2. Simulate long reads using SimLoRD while sampling lengths from a real PacBio long reads FASTQ file.\n",
    "\n",
    "The following commands for SimLoRD were used:\n",
    "\n",
    "```\n",
    "$simlord -n $reads -sf $fastq -rr $ref $outputPrefix -mp $passes\n",
    "```\n",
    "\n",
    "An explanation of the parameters used follows:\n",
    "* `-mp $passes` indicates that the maximum number of passes allowed for a given read. `$passes` was set to 1 for simulation of CLR data and 40 (which is the default for SimLoRD) to simulate CCS data.\n",
    "* `-sf $fastq` indicates that the lengths were sampled from a real PacBio FASTQ file. In this case, the _E.coli_ PacBio file SRR1284073 was used.\n",
    "\n",
    "## Analyses of Simulated Datasets ##\n",
    "Length distribution histograms, length-identity scatterplots and length - error rate scatter plots were created for each type of dataset (i.e. CLR and CCS). The mean identities and total error rates were also collected.\n",
    "\n",
    "To construct the length - identity scatterplots, the reads were mapped onto the reference genome using BLASR and the outputted SAM files were coverted to [Multiple Alignment Format](https://genome.ucsc.edu/FAQ/FAQformat.html#format5) for analyses.\n",
    "\n",
    "To construct the length - error rate scatterplots, the SAM files outputted by SimLoRD were converted to [Multiple Alignment Format](https://genome.ucsc.edu/FAQ/FAQformat.html#format5) for further analyses.\n",
    "\n",
    "#### Definitions ####\n",
    "Given an alignment between reference sequence A and read B, the _identity_ of B is defined as the number of positions in the alignment such that the bases of A and B are equal.\n",
    "\n",
    "Given an alignment between reference sequence A and read B, the _error rate_ of B is the number of positions in the alignment such that the bases of A and B are not equal.\n",
    "\n",
    "Given a set of SimLoRD reads S, the _total error rate_ of S is defined as the total number of erroneous bases in S divided by the total number of bases in S. This number is given a priori in the SimLoRD FASTQ files.\n",
    "\n",
    "## CLR Analyses ##\n",
    "\n",
    "### Mean Identities ###\n",
    "\n",
    "|  Coverage   | Mean Identity |\n",
    "| :--------:  | :-----------: |\n",
    "| 10          | 2884          |\n",
    "| 20          | 2873          |\n",
    "| 50          | 2897          |\n",
    "| 75          | 2879          |\n",
    "\n",
    "### Error Rate ###\n",
    "\n",
    "| Coverage | Error Rate |\n",
    "| :------: | :--------: |\n",
    "| 10       | 0.165      |\n",
    "| 20       | 0.165      |\n",
    "| 50       | 0.165      |\n",
    "| 75       | 0.165      |\n",
    "\n",
    "### Length Distributions ###\n",
    "\n",
    "#### Coverage 10 ####\n",
    "![Coverage 10 Length Distribution](ecoli_clr/length_distributions/long-d10.png)\n",
    "\n",
    "#### Coverage 20 ####\n",
    "![Coverage 20 Length Distribution](ecoli_clr/length_distributions/long-d20.png)\n",
    "\n",
    "#### Coverage 50 ####\n",
    "![Coverage 50 Length Distribution](ecoli_clr/length_distributions/long-d50.png)\n",
    "\n",
    "#### Coverage 75 ####\n",
    "![Coverage 75 Length Distribution](ecoli_clr/length_distributions/long-d75.png)\n",
    "\n",
    "### Length - Identity Scatterplots ###\n",
    "\n",
    "#### Coverage 10 ####\n",
    "![Coverage 10 Length Identity Scatterplot](ecoli_clr/length_identity_scatters/ecoli-long-d10_length_identity_scatter.png)\n",
    "\n",
    "#### Coverage 20 ####\n",
    "![Coverage 20 Length Identity Scatterplot](ecoli_clr/length_identity_scatters/ecoli-long-d20_length_identity_scatter.png)\n",
    "\n",
    "#### Coverage 50 ####\n",
    "![Coverage 50 Length Identity Scatterplot](ecoli_clr/length_identity_scatters/ecoli-long-d50_length_identity_scatter.png)\n",
    "\n",
    "#### Coverage 75 ####\n",
    "![Coverage 75 Length Identity Scatterplot](ecoli_clr/length_identity_scatters/ecoli-long-d75_length_identity_scatter.png)\n",
    "\n",
    "### Length - Error Rate Scatterplots ###\n",
    "\n",
    "#### Coverage 10 ####\n",
    "![Coverage 10 Length - Error Rate Scatter](ecoli_clr/error_rate_scatter/ecoli-long-d10_error_rate_scatter.png)\n",
    "\n",
    "#### Coverage 20 ####\n",
    "![Coverage 20 Length - Error Rate Scatter](ecoli_clr/error_rate_scatter/ecoli-long-d20_error_rate_scatter.png)\n",
    "\n",
    "#### Coverage 50 ####\n",
    "![Coverage 50 Length - Error Rate Scatter](ecoli_clr/error_rate_scatter/ecoli-long-d50_error_rate_scatter.png)\n",
    "\n",
    "#### Coverage 75 ####\n",
    "![Coverage 75 Length - Error Rate Scatter](ecoli_clr/error_rate_scatter/ecoli-long-d75_error_rate_scatter.png)\n",
    "\n",
    "## CCS Analysis ##\n",
    "\n",
    "### Mean Identities ###\n",
    "\n",
    "|  Coverage   | Mean Identity |\n",
    "| :--------:  | :-----------: |\n",
    "| 10          | 2838          |\n",
    "| 20          | 2843          |\n",
    "| 50          | 2791          |\n",
    "| 75          | 2821          |\n",
    "\n",
    "### Error Rate ###\n",
    "\n",
    "| Coverage | Error Rate |\n",
    "| :------: | :--------: |\n",
    "| 10       | 0.096      |\n",
    "| 20       | 0.096      |\n",
    "| 50       | 0.095      |\n",
    "| 75       | 0.096      |\n",
    "\n",
    "### Length Distributions ###\n",
    "\n",
    "#### Coverage 10 ####\n",
    "![Coverage 10 Length Distribution](ecoli_ccs/length_distributions/long-d10.png)\n",
    "\n",
    "#### Coverage 20 ####\n",
    "![Coverage 20 Length Distribution](ecoli_ccs/length_distributions/long-d20.png)\n",
    "\n",
    "#### Coverage 50 ####\n",
    "![Coverage 50 Length Distribution](ecoli_ccs/length_distributions/long-d50.png)\n",
    "\n",
    "#### Coverage 75 ####\n",
    "![Coverage 75 Length Distribution](ecoli_ccs/length_distributions/long-d75.png)\n",
    "\n",
    "### Length - Error Rate Scatterplots ###\n",
    "\n",
    "#### Coverage 10 ####\n",
    "![Coverage 10 Length - Error Rate Scatter](ecoli_ccs/error_rate_scatter/ecoli_ccs-long-d10_error_rate_scatter.png)\n",
    "\n",
    "#### Coverage 20 ####\n",
    "![Coverage 20 Length - Error Rate Scatter](ecoli_ccs/error_rate_scatter/ecoli_ccs-long-d20_error_rate_scatter.png)\n",
    "\n",
    "#### Coverage 50 ####\n",
    "![Coverage 50 Length - Error Rate Scatter](ecoli_ccs/error_rate_scatter/ecoli_ccs-long-d50_error_rate_scatter.png)\n",
    "\n",
    "#### Coverage 75 ####\n",
    "![Coverage 75 Length - Error Rate Scatter](ecoli_ccs/error_rate_scatter/ecoli_ccs-long-d75_error_rate_scatter.png)\n",
    "\n",
    "### Length - Identity Scatterplots ###\n",
    "\n",
    "#### Coverage 10 ####\n",
    "![Coverage 10 Length Identity Scatterplot](ecoli_ccs/length_identity_scatters/ecoli-long-d10_length_identity_scatter.png)\n",
    "\n",
    "#### Coverage 20 ####\n",
    "![Coverage 20 Length Identity Scatterplot](ecoli_ccs/length_identity_scatters/ecoli-long-d20_length_identity_scatter.png)\n",
    "\n",
    "#### Coverage 50 ####\n",
    "![Coverage 50 Length Identity Scatterplot](ecoli_ccs/length_identity_scatters/ecoli-long-d50_length_identity_scatter.png)\n",
    "\n",
    "#### Coverage 75 ####\n",
    "![Coverage 75 Length Identity Scatterplot](ecoli_ccs/length_identity_scatters/ecoli-long-d75_length_identity_scatter.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [Root]",
   "language": "python",
   "name": "Python [Root]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
