{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**13 June 2016**\n",
    "\n",
    "For yeast, files under directories `short-single-dx` where x = 50, 100, 200 created with commands:\n",
    "```\n",
    "$ PREFIX=/global/scratch/seanla/Data/yeast\n",
    "$ OUTPUTDIR=$PREFIX/short-single-d200\n",
    "$ OUTPUT=$OUTPUTDIR/yeast-short-single-d200\n",
    "$ cd /home/seanla/Software/art_bin_GreatSmokyMountains/\n",
    "$ mkdir -p $OUTPUTDIR\n",
    "$ ./art_illumina -s -sam -i $PREFIX/saccharomyces-cerevisiae chromosome1_sequence.fasta -l 100 -f x -o $OUTPUT\n",
    "```\n",
    "where x = 50, 100, 200."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For E.coli, files under directories `short-single-dx` where x = 50, 100, 200 created with commands:\n",
    "```\n",
    "$ PREFIX=/global/scratch/seanla/Data/ecoli\n",
    "$ OUTPUTDIR=$PREFIX/short-single-d200\n",
    "$ OUTPUT=$OUTPUTDIR/ecoli-short-single-d200\n",
    "$ cd /home/seanla/Software/art_bin_GreatSmokyMountains/\n",
    "$ mkdir -p $OUTPUTDIR\n",
    "$ ./art_illumina -s -sam -i $PREFIX/escherichia-coli_reference.fasta -l 100 -f x -o $OUTPUT\n",
    "```\n",
    "where x = 50, 100, 200\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**16 June 2016**\n",
    "\n",
    "Simulated long read files for E.coli with coverage of 10, 20, 50 and 75 were generated by SimLoRD using a shell/PBS script with the commands\n",
    "\n",
    "```\n",
    "PREFIX=/global/scratch/seanla/Data/ecoli\n",
    "PATH=/home/seanla/Projects/hcstats\n",
    "REF=$PREFIX/escherichia-coli_reference.fasta\n",
    "FASTQ=$PREFIX/SRR1284073.fastq\n",
    "\n",
    "\n",
    "COVERAGE=x\n",
    "READS=$(/global/software/python/python272/bin/python $PATH/reads4coverage.py -c $COVERAGE -i /global/scratch/seanla/Data/ecoli/SRR1284073.fastq -r /global/scratch/seanla/Data/ecoli/escherichia-coli_reference.fasta)\n",
    "OUTPUT=\"$PREFIX/simlord/long-d$COVERAGE/ecoli-long-d$COVERAGE\"\n",
    "\n",
    "/bin/mkdir -p $OUTPUT\n",
    "\n",
    "echo \"Coverage: $COVERAGE\"\n",
    "echo \"Reads: $READS\"\n",
    "\n",
    "/home/seanla/Software/anaconda2/envs/simlord/bin/simlord -n $READS -sf $FASTQ -rr $REF $OUTPUT\n",
    "```\n",
    "\n",
    "Where `x` was set to 10, 20, 50 and 75 respectively."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Simulated long read files for yeast with coverage of 10, 20, 50 and 75 were generated by SimLoRD using a shell/PBS script with the following commands:\n",
    "\n",
    "```\n",
    "# Paths to executables\n",
    "python=/global/software/python/python272/bin/python\n",
    "reads4coverage=/home/seanla/Projects/hcstats/reads4coverage.py\n",
    "simlord=/home/seanla/Software/anaconda2/envs/simlord/bin/simlord\n",
    "\n",
    "# Paths to data\n",
    "data=/global/scratch/seanla/Data/yeast\n",
    "ref=$data/saccharomyces-cerevisiae-chromosome1_sequence.fasta\n",
    "fastq=$data/SRR1284662.fastq\n",
    "\n",
    "# Perform simlord for coverage 10, 20, 50, 75\n",
    "\n",
    "# ------------------ Coverage x -----------------------------\n",
    "# Input parameters for simlord\n",
    "coverage=x\n",
    "# Find number of reads required for coverage\n",
    "reads=$($python $reads4coverage -c $coverage -i $fastq -r $ref)\n",
    "# Output directory path\n",
    "output=\"$data/simlord/long-d$coverage/ecoli-long-d$coverage\"\n",
    "\n",
    "mkdir -p $output\n",
    "\n",
    "echo \"Coverage: $coverage\"\n",
    "echo \"Reads: $reads\"\n",
    "\n",
    "$simlord -n $reads -sf $fastq -rr $ref $output\n",
    "```\n",
    "where `x` is 10, 20, 50 and 75."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "** 20 June 2016 **\n",
    "\n",
    "Paired-end short reads of coverage 50, 100 and 200 for ecoli and yeast were generated with the following commands:\n",
    "\n",
    "```\n",
    "prefix=/global/scratch/seanla/Data/ecoli\n",
    "outputdir=$prefix/art/short-paired-d200\n",
    "output=$outputdir/ecoli-short-paired-d200\n",
    "ref=$prefix/escherichia-coli_reference.fasta\n",
    "mkdir -p $outputdir\n",
    "cd /home/seanla/Software/art_bin_GreatSmokyMountains/\n",
    "./art_illumina -p -i $ref -l 100 -f x -m 300 -s 25 -o $output\n",
    "```\n",
    "\n",
    "```\n",
    "prefix=/global/scratch/seanla/Data/yeast\n",
    "outputdir=$prefix/art/short-paired-d100\n",
    "output=$outputdir/yeast-short-paired-d100\n",
    "ref=$prefix/accharomyces-cerevisiae-chromosome1_sequence.fasta\n",
    "mkdir -p $outputdir\n",
    "cd /home/seanla/Software/art_bin_GreatSmokyMountains/\n",
    "./art_illumina -p -i $ref -l 100 -f x -m 300 -s 25 -o $output\n",
    "```\n",
    "\n",
    "Where the `-l 100` indicates a short read length of 100, `-f x` indicates a fold coverage of `x` (of which `x` was set to 50, 100 and 200), `-m 300` indicates a mean fragment length of 300 and `-s 25` indicates a standard deviation of fragment length of 25."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "**21 June 2016**\n",
    "LoRDeC, Jabba and proovread were performed on short reads of coverage 50, 100 and 200 and long reads of coverage 10, 20, 50 and 75.\n",
    "\n",
    "For LoRDeC, the following parameters were used:\n",
    "- kmer length of 19\n",
    "- error rate of 0.4\n",
    "- number of target k-mers: 5\n",
    "- max branch exploration of 200\n",
    "\n",
    "For Jabba, the following parameters were used:\n",
    "- The de Bruijn graph was created with Brownie with a k-mer length of 45 (graphCorrection mode)\n",
    "- Thread usage of 6\n",
    "- Minimum size of MEM of 20\n",
    "- k-mer size of 45\n",
    "\n",
    "For proovread, preprocessing was performed as follows:\n",
    "- bwa-proovread was used to map the short reads onto the long reads\n",
    "- the resulting SAM file was sorted using samtools\n",
    "- For E.coli runs 50Sx50L, 50Sx75L, 100Sx50L, 100Sx75L, 200Sx10L, 200Sx20L, 200Sx50L, 200Sx75L, the parameter \"long read quality offset value\" could not be automatically determined and so the corrections failed. In future runs this value must be inputted manually.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Monday, 27 June 2016 ###\n",
    "\n",
    "LoRDeC, Jabba and proovread were used to correct long reads of coverage 10, 20, 50 and 75 using short reads of coverage 50, 100 and 200. Details of the pipeline are as follows:\n",
    "\n",
    "#### LoRDeC ####\n",
    "To run LoRDeC, the following command was used:\n",
    "\n",
    "```\n",
    "./lordec-correct -T ${PBS_NUM_PPN} --trials 5 --branch 200 --errorrate 0.4 -2 $short1 $short2 -k 19 -s 3 -i $long -o $output\n",
    "```\n",
    "\n",
    "An explanation of the parameters used are as follows:\n",
    "- `--trials` is the number of paths to try from a k-mer\n",
    "- `--branch` is the number of branches to explore\n",
    "- `--errorrate` is the expected maximum error rate of the long reads\n",
    "- `-k 19` is the k-mer length, which is set to 19\n",
    "- `-s 3` is the solid k-mer abundance threshold\n",
    "\n",
    "8 threads were allocated to each job.\n",
    "\n",
    "#### Jabba ####\n",
    "Short reads were first corrected with Karect with the following command:\n",
    "\n",
    "```\n",
    "karect -correct -inputfile=$short1 -inputfile=$short2 -resultdir=$output -tempdir=$output -celltype=haploid -matchtype=hamming -threads=${PBS_NUM_PPN}\n",
    "```\n",
    "\n",
    "The parameters are as follows:\n",
    "- Since we are working with E.coli and yeast genomes, the celltype was set to haploid.\n",
    "- We used the simple matching type of hamming.\n",
    "\n",
    "Next, we constructed the de Bruijn graph using the program Brownie using the following command:\n",
    "\n",
    "```\n",
    "brownie graphCorrection -k 75 -p $output $short1 $short2\n",
    "```\n",
    "\n",
    "Where k is the kmer length and was set to 75.\n",
    "\n",
    "Lastly, we corrected long reads using Jabba with the following command:\n",
    "\n",
    "```\n",
    "jabba -t ${PBS_NUM_PPN} -l 20 -k 75 -o $output -g $dbgraph -fastq $long\n",
    "```\n",
    "\n",
    "Where the parameters `l` and `k` are the minimum seed suze and kmer size, respectively. Notice that the kmer size of Jabba and Brownie are the same.\n",
    "8 threads were allocated to each job.\n",
    "\n",
    "#### proovread ####\n",
    "Before running proovread, short reads were mapped onto long reads using BWA-proovread, a special version of the conventional BWA mapper built for proovread.\n",
    "\n",
    "First, we indexed the long reads with the command:\n",
    "\n",
    "```\n",
    "bwa-proovread index $long\n",
    "```\n",
    "\n",
    "Then we mapped the short reads onto the long reads with the command:\n",
    "\n",
    "```\n",
    "bwa-proovread mem $long $short1 $short2 > $samOutput\n",
    "```\n",
    "\n",
    "Then the outputted SAM file was sorted using SAM tools and converted into binary:\n",
    "\n",
    "```\n",
    "samtools sort -T $tempPath -o $bamOutput $samOutput\n",
    "```\n",
    "\n",
    "Finally, the long reads were corrected with proovread using the following command:\n",
    "\n",
    "```\n",
    "proovread -t ${PBS_NUM_PPN} --lr-qv-offset 70 --bam $bamOutput -l $long -p $output\n",
    "```\n",
    "\n",
    "The parameter `--lr-qv-offset 70` indicates that the expected long read phred quality offset is manually set to 70. Normally, proovread automatically detects the quality offset of the long reads, but in datasets where the quality of the long reads are between 64 and 73 this fails so the quality offset must be set manually. The proovread developer indicated through correspondence on Github that this quality offsets for long reads are actually not needed for long reads, so we set the offset to an arbitrary value of 70. 8 threads were allocated to each job.\n",
    "\n",
    "## Wesdnesday, 6 July 2016 ##\n",
    "\n",
    "#### CoLoRMap ####\n",
    "CoLoRMap was used to correct long reads of coverage 10, 20, 50 and 75 using short reads of coverage 50, 100 and 200 with commands similar to:\n",
    "\n",
    "```bash\n",
    "$colormap $long $mergedShort $outputPrefix ${PBS_NUM_PPN}\n",
    "```\n",
    "\n",
    "The script `runBoth.sh` calls both the `runCorr.sh` and the `runOEA.sh` script to perform regular corrections and corrections with one-end anchors, respectively. 8 threads were allocated to each job."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
